[libSyntax] Allow adding garbage nodes in between any two children of a syntax node

ahoppen opened the pull request:
When the source code is invalid, this allows us to represent tokens that could not be used to form a valid syntax tree with more fidelity.

This commit does not start using GarbageNodes yet, it just sets everything up for them.

ahoppen commented:
https://github.com/apple/swift-syntax/pull/514

@swift-ci Please smoke test

rintaro reviewed (commented):
Before proceeding this, I think we should define the semantics of "garbage" nodes for traversal. Like, should `SyntaxVisitor` visit the contents of garbages, should `TokenSequence` includes tokens in garbages? How about `next/previousSibling`, `first/lastToken` etc.?

ahoppen commented on utils/gyb_syntax_support/CommonNodes.py:
```diff
@@ -51,4 +51,10 @@
                    requires_leading_newline=True),
          ]),
+
+    Node('GarbageNodes', kind='SyntaxCollection', element='Syntax',
```
OK, this is a bigger topic about how traversal should handle garbage and missing nodes. I believe there are two distinctly use cases for the SwiftSyntax tree
1. For use cases like a formatter, you want a fully source accurate representation of the syntax tree that includes all garbage but skips over missing nodes
2. If you are doing lightweight structural analysis of a Swift file, you want the parser to “fix-up” the syntax tree as much as possible using it’s heuristics, giving you a view of the tree that’s at least valid with respect to SwiftSyntax’s grammar. This is done by ignoring garbage nodes but walking missing tokens as if they were present.

I think these are both valid use case and sufficiently different that we shouldn’t make a decision which view should be the default. I implemented the two different views in https://github.com/apple/swift-syntax/pull/522.

Regarding the need for the `GarbageNodes` type: Theoretically, we could just use `SyntaxCollection` here, but IIRC the entire code generation pipeline is not properly set up to instantiate nodes of these base types. The `GarbageNodes` type also comes in handy for the fixed up view because it allows us to skip over all children in `GarbageSyntax`.

ahoppen commented on utils/gyb_syntax_support/Node.py:
```diff
@@ -22,7 +23,24 @@ def __init__(self, name, description=None, kind=None, traits=None,
+        # Add implicitly generated GarbageNodes children in between any two 
+        # defined children
+        if kind != 'SyntaxCollection':
+            for i in range(2 * len(children) + 1):
```
I agree that the trailing garbage is of very little value because most likely the current node doesn’t know until which node skip garbage tokens. I reduced the number of nodes to 2n.

Note that the placement of garbage nodes is still ambiguous because placement of nodes in `[garbage root 2]` and `[garbage bar 1]` results in the same syntax tree. I doubt this will be a problem in practice though. Whichever node knows that it wants to skip over tokens should place these in its garbage.

ahoppen commented on utils/gyb_syntax_support/Node.py:
```diff
@@ -22,7 +23,24 @@ def __init__(self, name, description=None, kind=None, traits=None,
+        self.children = []
+        # Add implicitly generated GarbageNodes children in between any two 
+        # defined children
+        if kind != 'SyntaxCollection':
```
My idea was that the elements of the collection contain the garbage. Most likely the element parsing functions will also be the ones that know how much garbage to skip.

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
https://github.com/apple/swift-syntax/pull/514

@swift-ci Please smoke test

ahoppen committed:
[libSyntax] Allow adding garbage nodes in between any two children of a syntax node

When the source code is invalid, this allows us to represent tokens that could not be used to form a valid syntax tree with more fidelity.

This commit does not start using GarbageNodes yet, it just sets everything up for them.

ahoppen commented:
https://github.com/apple/swift-syntax/pull/514

@swift-ci Please smoke test

ahoppen commented:
https://github.com/apple/swift-syntax/pull/514

@swift-ci Please smoke test

ahoppen commented:
https://github.com/apple/swift-syntax/pull/514

@swift-ci Please smoke test

ahoppen commented:
https://github.com/apple/swift-syntax/pull/514

@swift-ci Please smoke test

ahoppen commented:
https://github.com/apple/swift-syntax/pull/514

@swift-ci Please smoke test

ahoppen commented:
https://github.com/apple/swift-syntax/pull/514

@swift-ci Please smoke test
